db/batch/batch_impl.go

```diff
@@ -93,22 +93,39 @@ func (b *baseKVStoreBatch) Entry(index int) (*WriteInfo, error) {
 }
 
 func (b *baseKVStoreBatch) SerializeQueue(serialize WriteInfoSerialize, filter WriteInfoFilter) []byte {
-	b.mutex.Lock()
-	defer b.mutex.Unlock()
-	// 1. This could be improved by being processed in parallel
-	// 2. Digest could be replaced by merkle root if we need proof
-	bytes := make([]byte, 0)
-	for _, wi := range b.writeQueue {
-		if filter != nil && filter(wi) {
-			continue
-		}
-		if serialize != nil {
-			bytes = append(bytes, serialize(wi)...)
-		} else {
-			bytes = append(bytes, wi.Serialize()...)
-		}
+	// 1. Digest could be replaced by merkle root if we need proof
```

Comment:
```
b.mutex.Lock()
defer b.mutex.Unlock()
```
should the mtx be kept?

Reply:
Hey @Liuhaai , @envestcc  said it's not needed as the concurrent writes now happen on different indices within the slice

Reply:
https://stackoverflow.com/questions/42238425/golang-mutex-ranging-over-shared-array-in-goroutines

Reply:
could write an unit test to verify the concurrency problem

Reply:
mtx is for `b.writeQueue`

Reply:
Added the mtx back @Liuhaai

Reply:
does the `Unlock` can be called earlier? It seems like that it can be placed just before `wg.Wait()`.

but it seems that this is unrelated to the current issue. It may not improve the serialization process. Perhaps it can be discussed in a new issue.

Reply:
does the `Unlock` can be called earlier? It seems like that it can be placed just before `wg.Wait()`.

but it seems that this is unrelated to the current issue. It may not improve the serialization process. Perhaps it can be discussed in a new issue.